File: 3kyu/Battleship_field_validator.py
# -*- coding: utf-8 -*-

"""
Before the game begins, players set up the board and place the ships accordingly to the following rules:
- There must be single battleship (size of 4 cells), 2 cruisers (size 3), 3 destroyers (size 2) and 4 submarines (size 1). Any additional ships are not allowed, as well as missing ships.
- Each ship must be a straight line, except for submarines, which are just single cell.
- The ship cannot overlap or be in contact with any other ship, neither by edge nor by corner.
"""

# broken logic
def validate_battlefield(battleField):
    total_cell = 20
    positioning = []

    for grid_idx, grid in enumerate(battleField):
        for idx, value in enumerate(grid):
            if value:
                positioning.append((grid_idx, idx))

    if total_cell != len(positioning):
        return False

    row = {'row': True}
    col = {'col': True}
    idx = 0
    while idx < len(positioning):
        if not positioning:
            break
        if idx == -1:
            idx = 0

        cell = positioning[idx]
        # Returns false if it is an edge.
        if (cell[0] - 1, cell[1] - 1) in positioning:
            return False
        if (cell[0] + 1, cell[1] - 1) in positioning:
            return False
        if (cell[0] - 1, cell[1] + 1) in positioning:
            return False
        if (cell[0] + 1, cell[1] + 1) in positioning:
            return False

        if (cell[0] + 1, cell[1]) in positioning:
            if col.get(cell[1]):
                col[cell[1]].append(cell)
            else:
                col[cell[1]] = [cell]

            if not ((cell[0] + 2, cell[1]) in positioning):
                col[cell[1]].append((cell[0] + 1, cell[1]))
                positioning.remove((cell[0] + 1, cell[1]))

            positioning.pop(idx)
            idx -= 1
        elif (cell[0], cell[1] + 1) in positioning:
            if row.get(cell[0]):
                row[cell[0]].append(cell)
            else:
                row[cell[0]] = [cell]

            if not ((cell[0], cell[1] + 2) in positioning):
                row[cell[0]].append((cell[0], cell[1] + 1))
                positioning.remove((cell[0], cell[1] + 1))

            positioning.pop(idx)
            idx -= 1
        else:
            idx += 1

    number_of_submarines_found = len(positioning)
    print(col, '/', row, '/', positioning)
    ships = {4: 1, 3: 2, 2: 3, 1: 4}

    def check_number_of_ship(data):
        row = data.get('row')
        col = data.get('col')

        if row:
            data.pop('row')
        if col:
            data.pop('col')

        for k, lt in data.items():
            executed = False
            ship_range = len(lt)

            tmp = 0
            is_first = True
            for i, v in enumerate(lt):
                if is_first:
                    is_first = False
                    continue

                if ship_range == i + 1 and not executed:
                    break
                if row:
                    if (v[0], v[1] - 1) == lt[i - 1]:
                        if i + 1 != len(lt):
                            if (v[0], v[1] + 1) == lt[i + 1]:
                                continue
                        executed = True
                        checked_ship_range = len(lt[tmp:i + 1])
                        if checked_ship_range:
                            tmp = i + 1
                            ships[checked_ship_range] = ships.get(checked_ship_range) - 1
                if col:
                    if (v[0] - 1, v[1]) == lt[i - 1]:
                        if i + 1 != len(lt):
                            if (v[0] + 1, v[1]) == lt[i + 1]:
                                continue
                        executed = True
                        checked_ship_range = len(lt[tmp:i + 1])
                        print(checked_ship_range)
                        if checked_ship_range:
                            tmp = i + 1
                            ships[checked_ship_range] = ships.get(checked_ship_range) - 1

                # 세로버전 제작 필요.
            if executed:
                continue

            if ship_range == 5:
                ship_range = ship_range - 1
                ships[1] -= 1
            ships[ship_range] = ships.get(ship_range) - 1

        return True

    row_result = check_number_of_ship(row)
    col_result = check_number_of_ship(col)
    ships[1] = ships.get(1) - number_of_submarines_found
    bat, cur, des, sub = ships.items()

    print(row_result, col_result, ships, (bat[1], cur[1], des[1], sub[1]) == (0, 0, 0, 0))
    if row_result and col_result and (bat[1], cur[1], des[1], sub[1]) == (0, 0, 0, 0):
        return True
    else:
        return False

if __name__ == "__main__":
    test1 = [[1, 0, 0, 0, 0, 1, 1, 0, 0, 0],
             [1, 0, 1, 0, 0, 0, 0, 0, 1, 0],
             [1, 0, 1, 0, 1, 1, 1, 0, 1, 0],
             [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
             [0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
             [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
             [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]] # this test is vaild.
    test2 = [[1, 0, 0, 0, 0, 1, 1, 0, 0, 0],
             [1, 1, 0, 0, 0, 0, 0, 0, 1, 0],
             [1, 1, 0, 0, 1, 1, 1, 0, 1, 0],
             [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
             [1, 0, 0, 0, 0, 0, 0, 0, 1, 0],
             [0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
             [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]

    test3 = [[0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
             [0, 0, 0, 0, 1, 0, 1, 0, 0, 0],
             [0, 0, 0, 0, 0, 0, 1, 0, 1, 1],
             [0, 0, 1, 0, 1, 0, 0, 0, 0, 0],
             [0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
             [0, 0, 0, 1, 0, 1, 1, 1, 1, 0],
             [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
             [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
             [0, 0, 0, 0, 1, 0, 1, 1, 1, 0],
             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]

    test4 = [[0, 0, 0, 1, 0, 1, 0, 0, 0, 0],
             [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0, 1, 1, 0, 0, 0],
             [0, 1, 0, 1, 0, 0, 0, 0, 0, 0],
             [0, 0, 0, 1, 0, 0, 0, 0, 0, 1],
             [0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
             [0, 1, 0, 0, 0, 0, 0, 1, 0, 0],
             [0, 1, 0, 1, 0, 0, 0, 1, 0, 0],
             [0, 0, 0, 1, 0, 0, 0, 1, 0, 0],
             [0, 0, 0, 1, 0, 1, 0, 1, 0, 0]]

    test5 = [[0, 1, 0, 0, 0, 0, 1, 0, 0, 0],
             [0, 1, 0, 0, 1, 0, 0, 0, 0, 0],
             [0, 0, 0, 0, 1, 0, 0, 0, 0, 1],
             [1, 1, 1, 0, 1, 0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0, 0, 1, 1, 1, 1],
             [0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
             [1, 0, 0, 0, 1, 0, 0, 0, 0, 0],
             [1, 0, 0, 0, 0, 0, 0, 1, 0, 1],
             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]

    test6 = [[1, 0, 0, 0, 0, 1, 1, 0, 0, 0],
             [1, 1, 0, 0, 0, 0, 0, 0, 1, 0],
             [1, 1, 0, 0, 1, 1, 1, 0, 1, 0],
             [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
             [1, 0, 0, 0, 0, 0, 0, 0, 1, 0],
             [0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
             [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
             [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]

    print(validate_battlefield(test3))

File: 3kyu/Battleship_field_validator.js
/*
Before the game begins, players set up the board and place the ships accordingly to the following rules:

There must be single battleship (size of 4 cells), 2 cruisers (size 3), 3 destroyers (size 2) and 4 submarines (size 1). Any additional ships are not allowed, as well as missing ships.
Each ship must be a straight line, except for submarines, which are just single cell.
The ship cannot overlap, but can be contact with any other ship.
*/

function validateBattlefield(field) {
  const maxX = field.length;
  const maxY = field[0].length;
  const queue = [];
  // Arrangement to navigate up, down, left and right, including yourself
  const dx = [0, -1, 1, 0, 0];
  const dy = [0, 0, 0, -1, 1];
  // Array for searching the location of a collision with a combat ship
  const sdx = [-1, -1, -1, 0, 0, 1, 1, 1];
  const sdy = [-1, 0, 1, -1, 1, -1, 0, 1];

  const ship = {
    battleship: 0,
    cruiser: 0,
    destroyer: 0,
    submarine: 0,
  };

  for (let x = 0; x < maxX; x++) {
    for (let y = 0; y < maxY; y++) {
      if ([0, 2].includes(field[x][y])) continue;
      if (field[x][y] === 1) queue.push([x, y]);
      let shipLocation = [];
      while (queue.length > 0) {
        const [m, n] = queue.shift();
        for (let i = 0; i < dx.length; i++) {
          const nx = m + dx[i];
          const ny = n + dy[i];
          if (nx < 0 || nx >= maxX || ny < 0 || ny >= maxY) continue;
          if (field[nx][ny] === 1) {
            field[nx][ny] = 2; // 방문 처리
            queue.push([nx, ny]);
            shipLocation.push([nx, ny]);
          }
        }
      }

      // validate the ship
      if (shipLocation.length > 1) {
        const xLocOfShip = shipLocation.map((ship) => ship[0]);
        const yLocOfShip = shipLocation.map((ship) => ship[1]);
        if (
          xLocOfShip.every((x) => x === xLocOfShip[0]) ===
          yLocOfShip.every((y) => y === yLocOfShip[0])
        )
          return false;
      }
      // Collision detection with other combat ships
      for (const loc of shipLocation) {
        for (let i = 0; i < sdx.length; i++) {
          const nx = loc[0] + sdx[i];
          const ny = loc[1] + sdy[i];
          if (shipLocation.some((a) => a[0] === nx && a[1] === ny)) continue;
          if (nx < 0 || nx >= maxX || ny < 0 || ny >= maxY) continue;
          if (field[nx][ny] === 1) return false;
        }
      }
      switch (shipLocation.length) {
        case 1:
          ship.submarine += 1;
          break;
        case 2:
          ship.destroyer += 1;
          break;
        case 3:
          ship.cruiser += 1;
          break;
        case 4:
          ship.battleship += 1;
          break;
      }
    }
  }

  if (
    ship.battleship === 1 &&
    ship.cruiser === 2 &&
    ship.destroyer === 3 &&
    ship.submarine === 4
  ) {
    return true;
  }
  return false;
}

const test1 = [
  [1, 0, 0, 0, 0, 1, 1, 0, 0, 0],
  [1, 0, 1, 0, 0, 0, 0, 0, 1, 0],
  [1, 0, 1, 0, 1, 1, 1, 0, 1, 0],
  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
  [0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
]; // true

const test2 = [
  [1, 0, 0, 0, 0, 1, 1, 0, 0, 0],
  [1, 1, 0, 0, 0, 0, 0, 0, 1, 0],
  [1, 1, 0, 0, 1, 1, 1, 0, 1, 0],
  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [1, 0, 0, 0, 0, 0, 0, 0, 1, 0],
  [0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
]; // false

const test3 = [
  [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
  [0, 0, 0, 0, 1, 0, 1, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 1, 0, 1, 1],
  [0, 0, 1, 0, 1, 0, 0, 0, 0, 0],
  [0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 1, 0, 1, 1, 1, 1, 0],
  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 1, 0, 1, 1, 1, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
]; // false

const test4 = [
  [0, 0, 0, 1, 0, 1, 0, 0, 0, 0],
  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 1, 1, 0, 0, 0],
  [0, 1, 0, 1, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 1, 0, 0, 0, 0, 0, 1],
  [0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 1, 0, 0, 0, 0, 0, 1, 0, 0],
  [0, 1, 0, 1, 0, 0, 0, 1, 0, 0],
  [0, 0, 0, 1, 0, 0, 0, 1, 0, 0],
  [0, 0, 0, 1, 0, 1, 0, 1, 0, 0],
]; // true

const test5 = [
  [0, 1, 0, 0, 0, 0, 1, 0, 0, 0],
  [0, 1, 0, 0, 1, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 1, 0, 0, 0, 0, 1],
  [1, 1, 1, 0, 1, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 1, 1, 1, 1],
  [0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
  [1, 0, 0, 0, 1, 0, 0, 0, 0, 0],
  [1, 0, 0, 0, 0, 0, 0, 1, 0, 1],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
]; // true

const test6 = [
  [1, 0, 0, 0, 0, 1, 1, 0, 0, 0],
  [1, 1, 0, 0, 0, 0, 0, 0, 1, 0],
  [1, 1, 0, 0, 1, 1, 1, 0, 1, 0],
  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [1, 0, 0, 0, 0, 0, 0, 0, 1, 0],
  [0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
]; // false

const test7 = [
  [1, 0, 0, 0, 0, 1, 1, 0, 0, 0],
  [1, 0, 1, 0, 0, 0, 0, 0, 1, 0],
  [1, 0, 1, 0, 1, 1, 1, 0, 1, 0],
  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
  [0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
  [0, 0, 0, 1, 0, 0, 0, 0, 1, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
]; // false

const test8 = [
  [1, 0, 0, 0, 0, 1, 1, 0, 0, 0],
  [1, 0, 0, 0, 0, 0, 0, 0, 1, 0],
  [1, 1, 0, 0, 1, 1, 1, 0, 1, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
  [0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
  [0, 1, 0, 1, 0, 0, 0, 0, 0, 0],
  [0, 1, 0, 0, 0, 0, 0, 1, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
];

console.log(validateBattlefield(test8));
